lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp

```diff
@@ -0,0 +1,142 @@
+  template <typename XOp, typename YOp, typename ZOp, class Op>
+  void replaceWithIntrinsicOrDeviceFunction(Op operation, LLVM::LLVMDialect *dialect,
+                            unsigned indexBitwidth) {
+    assert(operation.getType().isIndex() &&
+           "expected an operation returning index");
+    OpBuilder builder(operation);
+    auto loc = operation.getLoc();
+    Value *newOp;
+    switch (dimensionToIndex(operation)) {
+    case X:
+      newOp = builder.create<XOp>(loc, LLVM::LLVMType::getInt32Ty(dialect));
+      break;
+    case Y:
+      newOp = builder.create<YOp>(loc, LLVM::LLVMType::getInt32Ty(dialect));
+      break;
+    case Z:
+      newOp = builder.create<ZOp>(loc, LLVM::LLVMType::getInt32Ty(dialect));
+      break;
+    default:
+      operation.emitError("Illegal dimension: " + operation.dimension());
+      signalPassFailure();
+      return;
+    }
+
+    if (indexBitwidth > 32) {
+      newOp = builder.create<LLVM::SExtOp>(
+          loc, LLVM::LLVMType::getIntNTy(dialect, indexBitwidth), newOp);
+    } else if (indexBitwidth < 32) {
+      newOp = builder.create<LLVM::TruncOp>(
+          loc, LLVM::LLVMType::getIntNTy(dialect, indexBitwidth), newOp);
+    }
+    operation.replaceAllUsesWith(newOp);
+    operation.erase();
+  }
+
+public:
+  void runOnFunction() {
+    LLVM::LLVMDialect *llvmDialect =
+        getContext().getRegisteredDialect<LLVM::LLVMDialect>();
+    unsigned indexBitwidth =
+        llvmDialect->getLLVMModule().getDataLayout().getPointerSizeInBits();
+    getFunction().walk([&](Operation *opInst) {
+      if (auto threadId = dyn_cast<gpu::ThreadId>(opInst)) {
+        replaceWithIntrinsicOrDeviceFunction<ROCDL::ThreadIdXOp, ROCDL::ThreadIdYOp,
+                             ROCDL::ThreadIdZOp>(threadId, llvmDialect,
+                                                indexBitwidth);
+        return;
+      }
+      if (auto blockId = dyn_cast<gpu::BlockId>(opInst)) {
+        replaceWithIntrinsicOrDeviceFunction<ROCDL::BlockIdXOp, ROCDL::BlockIdYOp,
+                             ROCDL::BlockIdZOp>(blockId, llvmDialect,
+                                               indexBitwidth);
+        return;
+      }
+
+      // BlockDimX/Y/Z are implemented as device functions on ROCDL.
+      if (auto blockDim = dyn_cast<gpu::BlockDim>(opInst)) {
+        replaceWithIntrinsicOrDeviceFunction<
+            ROCDL::BlockDimXOp, ROCDL::BlockDimYOp, ROCDL::BlockDimZOp>(
+            blockDim, llvmDialect, indexBitwidth);
+        return;
+      }
+      // GridDimX/Y/Z are implemented as device functions on ROCDL.
+      if (auto gridDim = dyn_cast<gpu::GridDim>(opInst)) {
+        replaceWithIntrinsicOrDeviceFunction<
+            ROCDL::GridDimXOp, ROCDL::GridDimYOp, ROCDL::GridDimZOp>(
+            gridDim, llvmDialect, indexBitwidth);
+        return;
+      }
+    });
+  }
+};
+
+} // anonymous namespace
+
+FunctionPassBase *createLowerGpuOpsToROCDLOpsPass() {
```

Comment:
This function should be qualified with mlir:: and not within a namespace.

Reply:
Will revise.